### PR TITLE
[WIP] Implement a DeprecationDetector class

### PR DESCRIPTION
**Pull request by MarvinKlemp**
Status::
- [x] #37 
- [x] Implemented simple configuration format, currently generated from CLI
- [x] Implemented DeprecationDetector class
- [x] Removed Pimple
- [x] Initial version of output
- [x] Added output information #49, #41 
- [x] Removed EventDispatcher 
- [x] Improve Namespaces
- [x] Tests

### CODE REVIEW

File: src/DeprecationDetector/DeprecationDetector.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+     * @return Violation[]
+     *
+     * @throws \Exception
+     */
+    public function checkForDeprecations($sourceArg, $ruleSetArg)
+    {
+        $ruleSet = $this->ruleSetLoader->loadRuleSet($ruleSetArg);
+
+        // TODO: not hard coded
+        $lib = (is_dir($ruleSetArg) ? $ruleSetArg : realpath('vendor'));
```
Comment (slde-gorillaman) [inline src/DeprecationDetector/DeprecationDetector.php]: The vendor directory could be read from the composer.json if it is defined in there, else I think it is ok to have a default hard coded

File: src/Finder/ParsedPhpFileFinder.php [hunk 2]
```diff
@@ -44,14 +40,12 @@ public function getIterator()
...
-                    $this->parser->parseFile($file);
-                } catch (\PhpParser\Error $ex) {
-                    $raw = $ex->getRawMessage() . ' in file ' . $file;
-                    $ex->setRawMessage($raw);
-                    $this->parserErrors[] = $ex;
-                }
+            try {
+                $this->parser->parseFile($file);
+            } catch (\PhpParser\Error $ex) {
+                $raw = $ex->getRawMessage().' in file '.$file;
```
Comment (slde-gorillaman) [inline src/Finder/ParsedPhpFileFinder.php]: cs?

File: src/Console/Command/CheckCommand.php [hunk 3]
```diff
@@ -108,99 +109,33 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-        $methodFilterOption = $input->getOption('filter-method-calls');
-        $methodFilterSetting = explode(',', $methodFilterOption);
-
-        $filter = new ComposedViolationFilter(
-            array(
-                new MethodViolationFilter($methodFilterSetting),
-            )
-        );
+        /* @TODO move to Detector */
+        //$output->writeln(sprintf('<comment>There are %s deprecations:</comment>', count($violations)));
```
Comment (slde-flash) [inline src/Console/Command/CheckCommand.php]: todo

File: src/Violation/ViolationFilter/MethodViolationFilter.php [hunk 4]
```diff
@@ -18,6 +18,16 @@ public function __construct(array $filterArray)
     }
 
     /**
+     * @param string $filterConfig
+     *
+     * @return MethodViolationFilter
+     */
+    public static function fromString($filterConfig)
+    {
+        return new self(explode(',', $filterConfig));
```
Comment (slde-flash) [inline src/Violation/ViolationFilter/MethodViolationFilter.php]: i prefer `new static`

File: tests/DeprecationDetector/DeprecationDetectorTest.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace SensioLabs\DeprecationDetector\DeprecationDetector\Tests;
+
+use SensioLabs\DeprecationDetector\DeprecationDetector\DeprecationDetector;
+
+class DeprecationDetectorTest extends \PHPUnit_Framework_TestCase
+{
+    public function testClassIsInitializable()
+    {
+        $ruleSetLoader = $this->prophesize('SensioLabs\DeprecationDetector\RuleSet\Loader\LoaderInterface');
```
Comment (slde-flash) [inline tests/DeprecationDetector/DeprecationDetectorTest.php]: why not `LoaderInterface::class`?

Comment (slde-thespirit) [inline tests/DeprecationDetector/DeprecationDetectorTest.php]: it's a php 5.5 feature

File: src/Console/Command/CheckCommand.php (hunk 3 above)
Comment (MarvinKlemp) [inline src/Console/Command/CheckCommand.php]: WIP

File: src/Finder/ParsedPhpFileFinder.php (hunk 2 above)
Comment (MarvinKlemp) [inline src/Finder/ParsedPhpFileFinder.php]: I'll run the fixer later :)

File: src/DeprecationDetector/DeprecationDetector.php (hunk 1 above)
Comment (MarvinKlemp) [inline src/DeprecationDetector/DeprecationDetector.php]: I don't like it either, but this PR is already really big

File: src/Violation/ViolationFilter/MethodViolationFilter.php (hunk 4 above)
Comment (MarvinKlemp) [inline src/Violation/ViolationFilter/MethodViolationFilter.php]: :+1:

File: src/Finder/ParsedPhpFileFinder.php [hunk 6]
```diff
@@ -14,23 +15,27 @@ class ParsedPhpFileFinder extends Finder
     protected $parser;
 
     /**
+     * @var VerboseProgressOutput
+     */
+    protected $dispatcher;
```
Comment (MarvinKlemp) [inline src/Finder/ParsedPhpFileFinder.php]: rename property to $progressOutput

File: src/TypeGuessing/SymbolTable/Visitor/SymbolTableVariableResolverVisitor.php [hunk 7]
```diff
@@ -41,6 +41,7 @@ public function enterNode(Node $node)
     {
         if ($node instanceof Node\Stmt\ClassLike) {
             $this->table->enterScope(new TableScope(TableScope::CLASS_LIKE_SCOPE));
+            /* @TODO bugs undefined property namespacedName a*/
```
Comment (MarvinKlemp) [inline src/TypeGuessing/SymbolTable/Visitor/SymbolTableVariableResolverVisitor.php]: Remove and open an issue

File: src/RuleSet/DirectoryTraverser.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ * Class Traverser.
+ *
+ * @author Christopher Hertel <christopher.hertel@sensiolabs.de>
+ */
+class DirectoryTraverser
+{
+    /**
+     * @var DeprecationParser
+     */
+    private $finder;
```
Comment (MarvinKlemp) [inline src/RuleSet/DirectoryTraverser.php]: wrong type

File: src/RuleSet/DirectoryTraverser.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * @param string $path
+     *
+     * @return null|RuleSet
+     */
+    public function traverse($path)
+    {
+        /* @TODO remove $quiet from DirectoryTraverser */
```
Comment (MarvinKlemp) [inline src/RuleSet/DirectoryTraverser.php]: should be removed

File: src/DetectorFactory.php [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+     */
+    private $ancestorResolver;
+
+    /**
+     * @param Configuration   $configuration
+     * @param OutputInterface $output
+     *
+     * @return DeprecationDetector
+     */
+    public function buildDetector(Configuration $configuration, OutputInterface $output)
```
Comment (MarvinKlemp) [inline src/DetectorFactory.php]: rename to `create()`

File: src/Configuration/Configuration.php [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        return $this->containerPath;
+    }
+
+    /**
+     * @return bool
+     */
+    public function useCachedRuleSet()
+    {
+        return $this->noCache;
```
Comment (slde-flash) [inline src/Configuration/Configuration.php]: rename to `$this->useCachedRuleSet`

File: src/Configuration/Configuration.php [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * @return string
+     */
+    public function ruleSetCacheDir()
+    {
+        return $this->cacheDir;
```
Comment (slde-flash) [inline src/Configuration/Configuration.php]: rename to `$this->ruleSetCacheDir`

File: src/Configuration/Configuration.php [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        return $this->filterMethodCalls;
+    }
+
+    /**
+     * @return bool
+     */
+    public function failOnDeprecation()
+    {
+        return $this->fail;
```
Comment (slde-flash) [inline src/Configuration/Configuration.php]: rename to `$this->failOnDeprecation`

File: src/Console/Command/CheckCommand.php [hunk 14]
```diff
@@ -32,7 +29,7 @@ protected function configure()
                         'composer.lock'
                     ),
                     new InputOption(
-                        'container-cache',
+                        'container-path',
```
Comment (slde-flash) [inline src/Console/Command/CheckCommand.php]: :-1: BC

File: src/Console/Command/CheckCommand.php [hunk 15]
```diff
@@ -108,99 +104,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-            $container['ruleset.cache']->disable();
-        } else {
-            $container['ruleset.cache']->setCacheDir($input->getOption('cache-dir'));
-        }
-
-        if ($input->getOption('verbose')) {
-            $container['event_dispatcher']->addSubscriber(new ProgressListener($output));
-        }
+        $factory = new DetectorFactory();
+        $detector = $factory->create($config, $output);
```
Comment (slde-flash) [inline src/Console/Command/CheckCommand.php]: `(new DetectorFactory())->create($config, $output);`

Comment (slde-thespirit) [inline src/Console/Command/CheckCommand.php]: not php 5.3 compatible

File: src/Console/Output/DefaultProgressOutput.php [hunk 16]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace SensioLabs\DeprecationDetector\Console\Output;
+
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class DefaultProgressOutput
+{
+    const PROGRESS_NAME = 'check';
```
Comment (slde-flash) [inline src/Console/Output/DefaultProgressOutput.php]: :+1: for PROCESS

File: src/Console/Output/DefaultProgressOutput.php [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * @var Stopwatch
+     */
+    private $stopwatch;
+
+    /**
+     * @param OutputInterface $output
+     * @param StopWatch       $stopwatch
+     */
+    public function __construct(OutputInterface $output, Stopwatch $stopwatch)
```
Comment (slde-flash) [inline src/Console/Output/DefaultProgressOutput.php]: Hide the Stopwatch using a decorator

File: src/DeprecationDetector.php [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * @param LoaderInterface       $ruleSetLoader
+     * @param AncestorResolver      $ancestorResolver
+     * @param ParsedPhpFileFinder   $deprecationFinder
+     * @param ViolationDetector     $violationDetector
+     * @param RendererInterface     $renderer
+     * @param DefaultProgressOutput $output
+     */
+    public function __construct(
+        LoaderInterface $ruleSetLoader,
```
Comment (slde-flash) [inline src/DeprecationDetector.php]: rename `LoaderInterface` to `RuleSetLoaderInterface`

File: src/DeprecationDetector.php [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        ViolationDetector $violationDetector,
+        RendererInterface $renderer,
+        DefaultProgressOutput $output
+    ) {
+        $this->ruleSetLoader = $ruleSetLoader;
+        $this->ancestorResolver = $ancestorResolver;
+        $this->deprecationFinder = $deprecationFinder;
+        $this->violationDetector = $violationDetector;
+        $this->renderer = $renderer;
+        $this->output = $output;
```
Comment (slde-flash) [inline src/DeprecationDetector.php]: avoid using the `output` "as a service".

File: src/DeprecationDetector.php [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * @param string $sourceArg
+     * @param string $ruleSetArg
+     *
+     * @return Violation[]
+     *
+     * @throws \Exception
+     */
+    public function checkForDeprecations($sourceArg, $ruleSetArg)
```
Comment (slde-flash) [inline src/DeprecationDetector.php]: inject the output here

File: src/RuleSet/DirectoryTraverser.php [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+use SensioLabs\DeprecationDetector\FileInfo\PhpFileInfo;
+use SensioLabs\DeprecationDetector\Finder\ParsedPhpFileFinder;
+use SensioLabs\DeprecationDetector\Parser\DeprecationParser;
+
+/**
+ * Class Traverser.
+ *
+ * @author Christopher Hertel <christopher.hertel@sensiolabs.de>
+ */
+class DirectoryTraverser
```
Comment (slde-flash) [inline src/RuleSet/DirectoryTraverser.php]: :+1: rename to `RuleSetDirectoryTraverser`

File: src/RuleSet/Loader/CouldNotLoadRuleSetException.php [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+
+class CouldNotLoadRuleSetException extends \Exception
+{
+    /**
+     * @param string $message
+     */
+    public function __construct($message)
+    {
+        parent::__construct(
+            sprintf('<error>%s</error>', $message)
```
Comment (slde-flash) [inline src/RuleSet/Loader/CouldNotLoadRuleSetException.php]: exception should not know about formatting

File: src/Violation/ViolationFilter/ComposedViolationFilter.php [hunk 23]
```diff
@@ -20,6 +20,14 @@ public function __construct(array $violationFilters)
     }
 
     /**
+     * @param ViolationFilterInterface $violationFilter
+     */
+    public function addViolationFilter(ViolationFilterInterface $violationFilter)
```
Comment (slde-flash) [inline src/Violation/ViolationFilter/ComposedViolationFilter.php]: just use constructor injection

File: src/Violation/ViolationFilter/MethodViolationFilter.php [hunk 24]
```diff
@@ -18,6 +18,16 @@ public function __construct(array $filterArray)
     }
 
     /**
+     * @param string $filterConfig
+     *
+     * @return MethodViolationFilter
+     */
+    public static function fromString($filterConfig)
+    {
+        return new static(explode(',', $filterConfig));
```
Comment (slde-flash) [inline src/Violation/ViolationFilter/MethodViolationFilter.php]: array_map(explode(',', $filterConfig), 'trim')

File: tests/DetectorFactoryTest.php [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace SensioLabs\DeprecationDetector\tests;
+
+use SensioLabs\DeprecationDetector\DetectorFactory;
+
+class DetectorFactoryTest extends \PHPUnit_Framework_TestCase
+{
+    public function testClassIsInitializable()
+    {
+        $factory = new DetectorFactory();
```
Comment (slde-flash) [inline tests/DetectorFactoryTest.php]: :-1: useless test?

File: tests/Finder/ParsedPhpFileFinderTest.php [hunk 26]
```diff
@@ -8,7 +8,15 @@ class ParsedPhpFileFinderTest extends \PHPUnit_Framework_TestCase
 {
     public function testClassIsInitializable()
```
Comment (slde-flash) [inline tests/Finder/ParsedPhpFileFinderTest.php]: :-1: useless test?

File: tests/RuleSet/DirectoryTraverserTest.php [hunk 27]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace SensioLabs\DeprecationDetector\Tests\RuleSet;
+
+use SensioLabs\DeprecationDetector\RuleSet\DirectoryTraverser;
+
+class DirectoryTraverserTest extends \PHPUnit_Framework_TestCase
+{
+    public function testClassIsInitializable()
```
Comment (slde-flash) [inline tests/RuleSet/DirectoryTraverserTest.php]: :-1: useless test?

File: tests/DetectorFactoryTest.php (hunk 25 above)
Comment (MarvinKlemp) [inline tests/DetectorFactoryTest.php]: IMO it's okay to explicit test a constructor

File: tests/Finder/ParsedPhpFileFinderTest.php (hunk 26 above)
Comment (MarvinKlemp) [inline tests/Finder/ParsedPhpFileFinderTest.php]: IMO it's okay to explicit test a constructor

File: tests/RuleSet/DirectoryTraverserTest.php (hunk 27 above)
Comment (MarvinKlemp) [inline tests/RuleSet/DirectoryTraverserTest.php]: IMO it's okay to explicit test a constructor

File: src/RuleSet/DirectoryTraverser.php (hunk 21 above)
Comment (MarvinKlemp) [inline src/RuleSet/DirectoryTraverser.php]: IMO :-1: 
because it is already in the `RuleSet` namespace

File: src/DeprecationDetector.php (hunk 18 above)
Comment (MarvinKlemp) [inline src/DeprecationDetector.php]: IMO :-1: 
because it is already in the `RuleSet` namespace

File: src/DeprecationDetector.php (hunk 20 above)
Comment (DrSchimke) [inline src/DeprecationDetector.php]: :+1:

File: src/DeprecationDetector.php (hunk 19 above)
Comment (DrSchimke) [inline src/DeprecationDetector.php]: :+1:

File: src/Violation/Renderer/RendererInterface.php [hunk 28]
```diff
@@ -10,4 +11,9 @@
      * @param Violation[] $violations
      */
     public function renderViolations(array $violations);
+
+    /**
+     * @param Error[] $errors
+     */
+    public function renderParserErrors(array $errors);
```
Comment (MarvinKlemp) [inline src/Violation/Renderer/RendererInterface.php]: @xabbuh @datenbrille This PR adds a `renderParserErrors` method to the `RendererInterface`.
Any suggestions how to display parser errors in the html output format ?

**Pull request merged**